### DISCUSSION TITLE
General questions

### DISCUSSION THREAD

**Discussion by stamminator** [DarkWanderer/ClickHouse.Client#155] [Q&A]
1. The Bulk Insertion page says "Column names can be optionally provided if source data has less columns than target table". Is there an example of this? I'm assuming it uses something like `IEnumerable<(string, object)[]>` instead of `IEnumerable<object[]>`? It would be nice if there were an overload of this in the assembly to make it more obvious.
2. When inserting smaller quantities of rows more frequently, what is the recommended approach?
3. For other "standard" ADO.NET queries and commands, does [normal syntax](https://stackoverflow.com/questions/17150542/how-to-insert-a-c-sharp-list-to-database-using-dapper-net) apply, or are there any Clickhouse-specific caveats?

**Comment by DarkWanderer** [comment]
Hi @stamminator ,

1. The column names are provided using one of the overloads:
```
        public Task WriteToServerAsync(IEnumerable<object[]> rows);
        public Task WriteToServerAsync(IEnumerable<object[]> rows, IReadOnlyCollection<string> columns);
        public Task WriteToServerAsync(IEnumerable<object[]> rows, CancellationToken token);
        public async Task WriteToServerAsync(IEnumerable<object[]> rows, IReadOnlyCollection<string> columns, CancellationToken token);
```
 2. The recommended approach for frequent insertion of multiple records is to maintain a single instance of `ClickHouseBulkCopy` and an associated connection (preferably with sessions feature off), and feed the data as it comes into `WriteToServerAsync`. If the data comes very often (multiple times per second), I'd recommend utilizing TPL "Dataflow" pattern and `BufferBlock` more specifically

 3. The parameters have special ClickHouse syntax: `{{parameter_name:parameter_type}}`, e.g. `SELECT {{my_param:String}}`

**Reply by stamminator** [reply]
Thanks for the follow-up. Any reason for using double-braces? The syntax seems to be `{parName:Array(UInt16)}`.

Do you have any examples of a parameterized query using ADO `ClickHouseCommand`? The [listed examples](https://github.com/DarkWanderer/ClickHouse.Client/wiki/SQL-Parameters#examples) shows the basic syntax of what ClickHouse wants, but not how your API might be used to do it. Specifically...

- Why does ClickHouseDbParameter's ToString() method return `$"{ParameterName}:{Value}"` instead of `$"{ParameterName}:{ClickHouseType}"` as per `{<name>:<data type>}`? I see how the logic in `HttpParameterFormatter.Format()` says "use ClickHouseType if set, otherwise map using value", but if that's the case, what's the purpose of that .ToString()?
- What does it looks like to create a valid parameterized ClickHouse SQL string, pass it into the ClickHouseCommand, create the parameters, pass them into ClickHouseCommand.Parameters, and execute?

**Reply by stamminator** [reply]
Ultimately I got around these issues by using the bulk insert. However, in the event someone wants to construct their own parameterized `INSERT INTO` query, I believe my questions are still applicable.
